src/encoder.rs

```diff
@@ -944,6 +947,61 @@ impl<T: Pixel> fmt::Display for FrameInvariants<T> {
   }
 }
 
+#[derive(Debug, Clone)]
+pub struct EncoderStats {
+  /// Stores counts of each block size used in this frame
+  pub block_size_counts: BTreeMap<BlockSize, usize>,
+  /// Stores the number of skip blocks used in this frame
+  pub skip_block_count: usize,
+  /// Stores counts of each transform type used in this frame
+  pub tx_type_counts: BTreeMap<TxType, usize>,
+  /// Stores counts of each prediction mode used for luma in this frame
+  pub luma_pred_mode_counts: BTreeMap<PredictionMode, usize>,
+  /// Stores counts of each prediction mode used in this frame
+  pub chroma_pred_mode_counts: BTreeMap<PredictionMode, usize>,
+}
+
+impl Default for EncoderStats {
+  fn default() -> Self {
+    EncoderStats {
+      block_size_counts: BTreeMap::new(),
+      skip_block_count: 0,
+      tx_type_counts: BTreeMap::new(),
+      luma_pred_mode_counts: BTreeMap::new(),
+      chroma_pred_mode_counts: BTreeMap::new(),
+    }
+  }
+}
+
+impl Add<&Self> for EncoderStats {
+  type Output = Self;
+
+  fn add(self, rhs: &EncoderStats) -> Self::Output {
+    let mut new = self;
+    rhs.block_size_counts.iter().for_each(|(&k, &v)| {
+      *new.block_size_counts.entry(k).or_insert(0) += v;
+    });
+    rhs.chroma_pred_mode_counts.iter().for_each(|(&k, &v)| {
+      *new.chroma_pred_mode_counts.entry(k).or_insert(0) += v;
+    });
+    rhs.luma_pred_mode_counts.iter().for_each(|(&k, &v)| {
+      *new.luma_pred_mode_counts.entry(k).or_insert(0) += v;
+    });
+    rhs.tx_type_counts.iter().for_each(|(&k, &v)| {
+      *new.tx_type_counts.entry(k).or_insert(0) += v;
+    });
+    new.skip_block_count += rhs.skip_block_count;
+    new
+  }
+}
+
+impl AddAssign<&Self> for EncoderStats {
+  fn add_assign(&mut self, rhs: &EncoderStats) {
+    let copy = self.clone();
```

Comment:
I think this is usually done the other way around to avoid clones. You put logic in `AddAssign`, then in `Add` you do `self += rhs; self`